### ISSUE TITLE
Support Unsupervised Training with Queries and Documents Similar to Generative Pseudo Labeling

### ISSUE THREAD

**Issue by guenthermi** [jina-ai/finetuner#673]
We want to support training text embedding models in an unsupervised way based on queries and documents provided by a user without the need of relevance scores or labels. 

This should be done on basis of [GPL](https://pypi.org/project/gpl/) with some adjustments to make it work with queries provided by the user. Moreover, we want to support not only English but also multi-lingual model. 

The following sub tasks are necessary for this integration

- [x] Implement functions in inetuner-core to generate a dataset with triplets and margin values similar as done in the  [gpl package](https://pypi.org/project/gpl/) https://github.com/jina-ai/finetuner-core/issues/520
- [x] Implement functions in finetuner-core to train text encoding models using such a dataset https://github.com/jina-ai/finetuner-core/issues/474
- [x] Support data generation jobs in the API https://github.com/jina-ai/finetuner-api/issues/322
- [x] Support data geneation jobs in the client https://github.com/jina-ai/finetuner/issues/712
- [x] #717

**Comment by tehami02** [comment]
@guenthermi Hey I think this can be done in this way -
To integrate unsupervised training of text embedding models based on user-provided queries and documents without relevance scores or labels, we can follow the following steps:

 We can implement functions in the client to generate triplets and margin values based on the user-provided queries and documents. A triplet will consists of an anchor text, a positive text, and a negative text, where the anchor text will be similar to the positive text and dissimilar to the negative text. The margin value will be used to determine the distance between the anchor and negative texts and the anchor and positive texts.

 After generating triplets and margin values, we can use finetuner-core to train text encoding models using this dataset. We can implement functions in finetuner-core to train text encoding models using the triplets generated by the client. This will involve optimizing a loss function based on the distance between the anchor and positive texts and the distance between the anchor and negative texts.

 Finally, we can integrate multi-lingual text encoding models into finetuner-core to support multiple languages. We can do it by using pre-trained language models such as BERT, XLM, or multilingual models. These pre-trained models can be fine-tuned using the triplet dataset generated by the client. This will allow the models to learn language-specific features and generalize across languages.

**Comment by guenthermi** [comment]
Yes, we plan to do it in a similar way. The training is already supported, however, generating the triplets will not be done inside the clients, but rather as a compute job, since we need to do it with GPUs since it is too slow otherwise.

**Comment by tehami02** [comment]
@guenthermi I think performing the triplet generation as a compute job on a GPU cluster will be a good approach to improve the efficiency of the process and accelerate the training of the model.

I have joined slack, I am a beginner in ML/AI but I would love to try to develop it. Do you have any personal suggestions as to how should I start apart from documentations ?

**Comment by tehami02** [comment]
Thanks for the reply. I think for now I will look into some open source content and learn more about how things work. All the best.
